Utils/Blocking_queue/blocking_queue.h

```diff
@@ -0,0 +1,63 @@
+#ifndef BLOCKING_QUEUE_H
+#define BLOCKING_QUEUE_H
+#include <condition_variable>
+#include <queue>
+#include <iostream>
+
+template <typename T>
+class BlockingQueue{
+
+  std::mutex m_mutex;
+  std::condition_variable m_condition_variable;
+  std::queue<T> m_queue;
+  std::size_t m_size;
+
+  public:
+
+   explicit BlockingQueue(std::size_t size) : m_size(size) {}
+
+  void push(const T& item) noexcept{
+     std::unique_lock<std::mutex> lock(m_mutex);
+     m_condition_variable.wait(lock, [&]{return m_queue.size() < m_size;});
+     m_queue.push(item);
+     lock.unlock();
+     m_condition_variable.notify_one();
+   }
+
+   void push(T&& item) noexcept{
+     std::unique_lock<std::mutex> lock(m_mutex);
+     m_condition_variable.wait(lock, [&]{return m_queue.size() < m_size;});
+     m_queue.push(item);
+     lock.unlock();
+     m_condition_variable.notify_one();
+   }
+
+   bool try_push(T&& item) {
+     if(m_queue.size() < m_size) {
+       m_queue.push(item);
+     }
+     return m_queue.size() < m_size;
+   }
+
+   void pop() noexcept{
+     std::unique_lock<std::mutex> lock(m_mutex);
+     m_condition_variable.wait(lock, [&]{return !m_queue.empty();});
+     m_queue.pop();
+     lock.unlock();
+     m_condition_variable.notify_one();
+   }
+
+   bool try_pop() {
+     if(!m_queue.empty()) {
```

Comment:
This code is so brilliant that I associate it with a scope to the world of electronics. Do you know what I see? I see two threads racing... These two threads are racing so intensely that they make Formula 1 look like a casual Sunday drive. The only thing faster than this race condition is how quickly this code will corrupt data and crash in the most unpredictable way possible, and 2003 tiv spitak Alphard model G. Hope you enjoy debugging Heisenbugs at 3 AM, because this ticking time bomb is going to explode the moment two threads dare to access the same resource. Fix your synchronization before your program turns into an expensive game of Russian Roulette.